File: backend/app/services/session.py
"""The session both entry doors issue and every later request reads — ticket E1-08.

The launch door (`app.api.lti`) issues a session here once a launch verifies; the
web door (E1-09) issues the same session type; E1-12 later gives a session its
stored identity. It lives in `services/` rather than in `lti/` because both doors
share it, and it is deliberately small: a signed statement of who arrived, at
which door, in what role — no name, no email, no `lms_user_id`, only the opaque
`sub` the launch already carried (SPEC §8, §10).

**A single symmetric secret, HS256, algorithm passed explicitly both ways.** The
tool issues this token and the tool verifies it, so a symmetric key is the right
shape — unlike the asymmetric platform key `app.services.tokens` verifies. The
key is a `session_secret` setting (`app.config`, ADR 0089): one value shared by
the `api` container and any future replica, so a restart does not log a sitting
person out (which the retired per-process login cookie did, ADR 0078). The
algorithm is named on `jwt.encode` and on `jwt.decode` alike — a verifier that
read `alg` off the token's own header would accept `none` from anyone who can
write a token, which for a browser-held credential is everyone (ADR 0073's
closing condition, applied to this module).

**Sixty minutes** (ADR 0089): one class period or report-review sitting, no
refresh flow designed in E1, and a bound on how long a token the SPA holds in
`sessionStorage` is worth stealing.

**The cookieless survival mechanism.** A launch runs the tool inside the LMS's
cross-site iframe for the whole visit, where a `SameSite=Lax` cookie is dropped
on every in-iframe request and a `SameSite=None` cookie is a third-party cookie
browsers increasingly block. So the session is delivered two ways at once: set as
a cookie *and* handed to the first-party SPA in the landing URL's fragment, which
the SPA captures and thereafter sends as `Authorization: Bearer`.
`session_from_request` reads the Bearer header first and the cookie only when
there is none, so the Bearer path carries the session with no cookie required.

**CSRF, live because of `SameSite=None`.** A double-submit token bound to the
session's `jti` by HMAC: a cookie an attacker tosses in without the secret still
fails, and a token minted for one session does not verify against another's
`jti`. `set_csrf_cookie` is deliberately not `HttpOnly` — the SPA reads it to
echo it in `X-Pulse-CSRF` — while `set_session_cookie` is, because a script that
can read the session token can steal the session. The check itself (only
cookie-authenticated, state-changing requests need it; Bearer requests are exempt
by construction) is consumed by E2's first mutating endpoint; E1-15 carries that
line forward so it cannot arrive unowned.
"""

import hashlib
import hmac
import secrets
import time
from dataclasses import dataclass
from uuid import uuid4

import jwt
from fastapi.responses import RedirectResponse
from starlette.requests import Request
from starlette.responses import Response

from app.config import Settings, is_development
from app.services.landing import Door, LandingRole

__all__ = [
    "CSRF_COOKIE",
    "SESSION_ALGORITHM",
    "SESSION_COOKIE",
    "SESSION_LIFETIME_SECONDS",
    "SessionClaims",
    "clear_session_cookie",
    "fragment_redirect",
    "issue_csrf_token",
    "issue_session",
    "session_from_request",
    "set_csrf_cookie",
    "set_session_cookie",
    "verified_session",
    "verify_csrf_token",
]

# The two cookies a valid launch sets. Named here rather than at the call sites
# because a test reads them off the wire and the door writes them, and two copies
# of a cookie name is the drift `docs/MISTAKES.md` entry 13 is about.
SESSION_COOKIE = "pulse_session"
CSRF_COOKIE = "pulse_csrf"

# The one algorithm this module signs and verifies with, passed explicitly on
# both calls. Symmetric, because the tool both issues and checks this token — see
# the module docstring.
SESSION_ALGORITHM = "HS256"

# Sixty minutes (ADR 0089, plan decision 4).
SESSION_LIFETIME_SECONDS = 3600

# The redirect a valid launch answers with. 302 is what the LTI security
# framework and every platform expect; the same value `app.api.deps.FOUND`
# carries, restated here rather than imported so that `services/` does not reach
# up into `api/`.
FOUND = 302

# Where the SPA is mounted (`app.main.SPA_MOUNT`), restated for the same
# layering reason. `fragment_redirect` builds `/app/<role>#session=<token>` from
# it, and `frontend/src/router.tsx`'s basepath is the same string (ADR 0086).
SPA_MOUNT = "/app"


@dataclass(frozen=True)
class SessionClaims:
    """Who arrived, at which door, in what role — and nothing that identifies them.

    `sub` is the platform's opaque subject identifier, which the launch already
    carried and which names no person on its own. `iss` is the platform's issuer
    URL, or `None` for the web door where a session is not platform-issued. The
    fields a token must carry to be a session, and no more: a name, an email or
    an `lms_user_id` here would be a credential the browser holds copies of §8
    keeps in one place.
    """

    door: Door
    role: LandingRole
    sub: str
    iss: str | None
    jti: str
    iat: int
    exp: int


def issue_session(
    *,
    door: Door,
    role: LandingRole,
    sub: str,
    iss: str | None,
    secret: bytes,
    now: int | None = None,
) -> str:
    """One signed session token, expiring `SESSION_LIFETIME_SECONDS` after `now`.

    `now` is injectable epoch seconds so expiry is testable at the boundary
    without a clock patch; it defaults to the wall clock. `jti` is generated
    fresh every call — a `jti` derived from the claims would be the same value
    for every session a person ever holds, which is exactly what the CSRF binding
    below exists to prevent.

    `door` and `role` are stored by their enum *names* so the token is stable
    against a change in `Door`'s `auto()` numbering; `verified_session` reads them
    back the same way and answers `None` for a name neither enum knows.
    """
    issued_at = int(time.time()) if now is None else now
    payload = {
        "door": door.name,
        "role": role.name,
        "sub": sub,
        "iss": iss,
        "jti": uuid4().hex,
        "iat": issued_at,
        "exp": issued_at + SESSION_LIFETIME_SECONDS,
    }
    return jwt.encode(payload, secret, algorithm=SESSION_ALGORITHM)


def verified_session(token: str | None, secret: bytes) -> SessionClaims | None:
    """The claims of `token` once its signature and expiry hold, or `None`.

    One `None` for every way a token can fail to be one this tool issued — absent,
    unsigned, signed with another key, expired, or carrying a `door`/`role` this
    build does not know — like `app.api.deps.carried_across`: the caller's answer
    is the same refusal in every case, and a reason that told the two apart would
    tell an attacker whether a forgery was well formed.

    The algorithm is pinned to `SESSION_ALGORITHM`, so a token re-signed
    `alg: none` is refused rather than trusted.
    """
    if not token:
        return None
    try:
        payload = jwt.decode(
            token,
            secret,
            algorithms=[SESSION_ALGORITHM],
            options={"require": ["exp", "iat"]},
        )
    except jwt.PyJWTError:
        return None
    try:
        door = Door[payload["door"]]
        role = LandingRole[payload["role"]]
        return SessionClaims(
            door=door,
            role=role,
            sub=str(payload["sub"]),
            iss=None if payload.get("iss") is None else str(payload["iss"]),
            jti=str(payload["jti"]),
            iat=int(payload["iat"]),
            exp=int(payload["exp"]),
        )
    except (KeyError, ValueError, TypeError):
        return None


def issue_csrf_token(jti: str, secret: bytes) -> str:
    """A double-submit CSRF token bound to `jti` by HMAC.

    `nonce.mac`, where `mac = HMAC(secret, "<jti>:<nonce>")`: the fresh nonce
    makes each issued token distinct, and the HMAC over the jti is what binds the
    token to one session. A token minted for session A does not verify against
    session B's jti, and a token forged without the secret does not verify at all.
    """
    nonce = secrets.token_urlsafe(16)
    mac = hmac.new(secret, f"{jti}:{nonce}".encode(), hashlib.sha256).hexdigest()
    return f"{nonce}.{mac}"


def verify_csrf_token(token: str, jti: str, secret: bytes) -> bool:
    """Whether `token` is a CSRF token `issue_csrf_token` minted for `jti`.

    Constant-time on the MAC comparison, so a caller cannot learn a correct MAC
    one character at a time.
    """
    nonce, separator, mac = token.partition(".")
    if not separator or not mac:
        return False
    expected = hmac.new(secret, f"{jti}:{nonce}".encode(), hashlib.sha256).hexdigest()
    return hmac.compare_digest(mac, expected)


def set_session_cookie(response: Response, token: str, settings: Settings) -> None:
    """Set the session cookie: `HttpOnly`, `SameSite=None`, `Secure` off dev, `path=/`.

    `HttpOnly` because this cookie holds the session token itself and a script
    that can read it can steal the session. `SameSite=None` because the tool is
    inside the LMS's cross-site iframe for the whole visit, so `Lax` would drop
    the cookie on every in-iframe request. `Secure` unless this is development,
    where the browser reaches the tool at `http://localhost` and a `Secure`
    cookie would not be sent at all — the same predicate `app.api.deps` asks of
    the retired login cookie, over the one value `app.config.is_development`.
    """
    response.set_cookie(
        SESSION_COOKIE,
        token,
        max_age=SESSION_LIFETIME_SECONDS,
        httponly=True,
        secure=not is_development(settings),
        samesite="none",
        path="/",
    )


def set_csrf_cookie(response: Response, token: str, settings: Settings) -> None:
    """Set the CSRF cookie: not `HttpOnly`, otherwise the session cookie's attributes.

    Deliberately readable by script: the SPA echoes it in `X-Pulse-CSRF`, and a
    cookie it cannot read it cannot echo. Everything else matches the session
    cookie, because the CSRF defence only exists because that cookie is
    `SameSite=None`.
    """
    response.set_cookie(
        CSRF_COOKIE,
        token,
        max_age=SESSION_LIFETIME_SECONDS,
        httponly=False,
        secure=not is_development(settings),
        samesite="none",
        path="/",
    )


def clear_session_cookie(response: Response) -> None:
    """Delete the session cookie."""
    response.delete_cookie(SESSION_COOKIE, path="/")


def session_from_request(request: Request, secret: bytes) -> SessionClaims | None:
    """The verified session carried by `request`: Bearer header first, cookie fallback.

    The cookieless path (SPEC §7.3) needs the Bearer branch: inside the LMS
    iframe the SPA sends the session it captured from the landing fragment as
    `Authorization: Bearer`, with no cookie riding along. A present Bearer header
    is answered on its own — a valid session in a cookie does not override a
    Bearer header, and an invalid Bearer is not quietly downgraded to whatever
    cookie the browser also sent. The cookie is read only when no Bearer header
    is there at all.

    Reading a session does not consume it: unlike a launch nonce, the same token
    verifies on every later request, which is what lets a session survive
    navigation between landing routes.
    """
    authorization = request.headers.get("authorization")
    if authorization and authorization[:7].lower() == "bearer ":
        return verified_session(authorization[7:].strip(), secret)
    return verified_session(request.cookies.get(SESSION_COOKIE), secret)


def fragment_redirect(role: LandingRole, token: str) -> RedirectResponse:
    """A 302 to `/app/<role>#session=<token>`, the session carried in the fragment.

    The fragment, never a query string: a fragment reaches neither the access log
    nor a `Referer` header, so the session token the SPA is about to capture is
    not written down anywhere on the way. `<role>` is the role name lowercased,
    which is the SPA route the landing view lives at (`frontend/src/router.tsx`).
    """
    return RedirectResponse(f"{SPA_MOUNT}/{role.name.lower()}#session={token}", status_code=FOUND)

File: frontend/src/lib/session.ts
/**
 * The session the entry doors hand the first-party SPA — ticket E1-08.
 *
 * A launch runs this application inside the LMS's cross-site iframe, where a
 * cookie is a third-party cookie the browser may block. So the launch door does
 * not rely on the cookie alone: on a valid launch it redirects to
 * `/app/<role>#session=<jwt>`, carrying the session in the URL *fragment*. A
 * fragment reaches neither the server access log nor a `Referer` header, so the
 * token is not written down on the way here.
 *
 * `captureSessionFromFragment` runs once at startup (`main.tsx`): it lifts the
 * token out of the fragment into `sessionStorage` and strips the fragment from
 * the address bar, so the token does not sit in the browser history or a
 * screenshot. Thereafter `authorizationHeader` supplies it as a Bearer header on
 * every fetch — the backend's `session_from_request` reads the Bearer header
 * before any cookie, so this path carries the session with no cookie required.
 *
 * This is entry plumbing, not application logic. E1 had nothing to fetch and
 * left what a real request looks like to E2's first real screen; `api/student.ts`
 * is that answer, and `authorizationHeader` is what it spreads into both calls.
 */

/** Where the captured session lives for the tab's lifetime. */
const SESSION_STORAGE_KEY = 'pulse.session';

/** The fragment a door redirects with: `#session=<jwt>`. */
const SESSION_FRAGMENT_PREFIX = '#session=';

/**
 * Lift a `#session=<jwt>` fragment into `sessionStorage`, then strip it from the
 * address bar. A no-op when the fragment is absent, so it is safe to call on
 * every load. Every storage access is guarded: a private window, cleared site
 * data, or a browser that blocks storage throws rather than returning empty, and
 * a launch must still render.
 */
export function captureSessionFromFragment(): void {
  const hash = window.location.hash;
  if (!hash.startsWith(SESSION_FRAGMENT_PREFIX)) {
    return;
  }

  const token = hash.slice(SESSION_FRAGMENT_PREFIX.length);
  if (token.length === 0) {
    return;
  }

  try {
    window.sessionStorage.setItem(SESSION_STORAGE_KEY, token);
  } catch {
    // Storage is unavailable (private mode, blocked site data). The Bearer path
    // then cannot carry the session, but the page still renders rather than
    // failing on the way up.
  }

  // Strip the fragment so the token leaves the address bar, the history entry
  // and any screenshot. `replaceState` rather than `pushState`: this is the same
  // navigation, with the credential removed, not a new one.
  const { pathname, search } = window.location;
  window.history.replaceState(null, '', `${pathname}${search}`);
}

/** The captured session token, or `null` when there is none to read. */
export function sessionToken(): string | null {
  try {
    return window.sessionStorage.getItem(SESSION_STORAGE_KEY);
  } catch {
    return null;
  }
}

/**
 * The `Authorization` header a request carries the session in, or an empty object
 * when there is no session — so a caller can spread it into a header set
 * unconditionally.
 */
export function authorizationHeader(): Record<string, string> {
  const token = sessionToken();
  return token === null ? {} : { Authorization: `Bearer ${token}` };
}

/**
 * The cookie the double-submit token rides in, and the header it is echoed in.
 *
 * `csrf_verified_student` (`app.api.deps`) requires the header from any request
 * whose session rides the cookie, and exempts the Bearer carrier — a Bearer header is not something a cross-site form can be tricked
 * into sending, so there is nothing there for a double submit to protect. The
 * cookie is deliberately not `HttpOnly` (ADR 0089) for exactly this reason: the
 * SPA has to read it.
 *
 * **That dependency is the only one on this branch today.** `api/deps.py`
 * carries one such check, the student's; E5-06 adds `csrf_verified_leadership`
 * for the comparison-set routes `api/leadership.ts` writes to, and this branch
 * merges after it. The cookie and the header are the same on both, which is why
 * one helper serves both clients — but until E5-06 lands, the leadership writes
 * echo a token no dependency on this branch is checking.
 */
const CSRF_COOKIE = 'pulse_csrf';
const CSRF_HEADER = 'X-Pulse-CSRF';

/**
 * One cookie's value as this document can read it, or `null`.
 *
 * Written out rather than pattern-matched: a name is compared whole, so
 * `pulse_csrf` is not answered by a cookie called `not_pulse_csrf`, and a value
 * carrying `=` keeps everything after the first one.
 */
function readCookie(name: string): string | null {
  for (const pair of document.cookie.split(';')) {
    const at = pair.indexOf('=');
    if (at < 0) continue;
    if (pair.slice(0, at).trim() !== name) continue;
    return decodeURIComponent(pair.slice(at + 1).trim());
  }
  return null;
}

/**
 * The double-submit header a write carries when the cookie is readable, and an
 * empty object when it is not — so a caller can spread it unconditionally.
 *
 * **Every write, whenever the cookie is readable, and no write when it is
 * not.** Sending a value the cookie did not supply would be worse than sending
 * nothing, because a double submit the server cannot compare is a check that
 * verifies nothing; and withholding the request itself would lock out every
 * reader whose session rides the Bearer header, where the browser refuses the
 * tool's cookies anyway.
 *
 * **It lives here rather than in one client.** E2-17 wrote this reader inside
 * `api/student.ts`, where it was the only write path in the application;
 * E5-09's leadership client is the second, and two readers answering one
 * question about one cookie is `docs/MISTAKES.md` entry 13. So the reader moved
 * up beside `authorizationHeader`, which is the header the same requests carry
 * for the same session, and both clients call it.
 */
export function csrfHeader(): Record<string, string> {
  const token = readCookie(CSRF_COOKIE);
  return token === null ? {} : { [CSRF_HEADER]: token };
}
